File: apps/api/app/modules/core/excel_filename.py
"""Tên file Excel xuất ra.

Lương / OT / KPI / chu kỳ: tháng.năm kỳ đang xuất
  → Lương 08.2026 công ty Dongju Sports VN.xlsx
Danh sách NV: ngày bấm xuất (giờ VN)
  → Danh sách nhân viên 20.08.2026 công ty Dongju Sports VN.xlsx
Windows không cho dấu / nên dùng chấm.
"""

from __future__ import annotations

from datetime import date, datetime, timedelta, timezone
from urllib.parse import quote

COMPANY_SUFFIX = "công ty Dongju Sports VN"
_VN = timezone(timedelta(hours=7))

_VI_FOLD = str.maketrans(
    "áàảãạăắằẳẵặâấầẩẫậéèẻẽẹêếềểễệíìỉĩịóòỏõọôốồổỗộơớờởỡợúùủũụưứừửữựýỳỷỹỵđ"
    "ÁÀẢÃẠĂẮẰẲẴẶÂẤẦẨẪẬÉÈẺẼẸÊẾỀỂỄỆÍÌỈĨỊÓÒỎÕỌÔỐỒỔỖỘƠỚỜỞỠỢÚÙỦŨỤƯỨỪỬỮỰÝỲỶỸỴĐ",
    "aaaaaaaaaaaaaaaaaeeeeeeeeeeeiiiiiooooooooooooooooouuuuuuuuuuuyyyyyd"
    "AAAAAAAAAAAAAAAAAEEEEEEEEEEEIIIIIOOOOOOOOOOOOOOOOOUUUUUUUUUUUYYYYYD",
)


def month_year_tag(period: str | None = None) -> str:
    """2026-08 → 08.2026. period=None → tháng hiện tại (VN)."""
    if period and period.strip():
        year_s, month_s, *_rest = f"{period.strip()}-01".split("-")
        return f"{int(month_s):02d}.{year_s}"
    now = datetime.now(_VN)
    return f"{now.month:02d}.{now.year}"


def export_day_tag(as_of: date | None = None) -> str:
    """20.08.2026 — ngày xuất hiện tại (VN)."""
    d = as_of or datetime.now(_VN).date()
    return d.strftime("%d.%m.%Y")


def company_excel_filename(
    label: str,
    *,
    period: str | None = None,
    extra: str | None = None,
    on_export_day: bool = False,
    as_of: date | None = None,
) -> str:
    stamp = export_day_tag(as_of) if (on_export_day or as_of is not None) else month_year_tag(period)
    parts = [label.strip(), stamp]
    if extra and extra.strip():
        parts.append(extra.strip())
    parts.append(COMPANY_SUFFIX)
    return f"{' '.join(parts)}.xlsx"


def ascii_filename_fallback(filename: str) -> str:
    folded = filename.translate(_VI_FOLD)
    safe = "".join(ch if ch.isascii() and ch not in '\\/:"*?<>|' else "_" for ch in folded)
    safe = "_".join(safe.split())
    return safe or "download.xlsx"


def attachment_content_disposition(filename: str) -> str:
    """RFC 5987 — trình duyệt lấy filename* (tiếng Việt), ASCII là dự phòng."""
    fallback = ascii_filename_fallback(filename)
    return f"attachment; filename=\"{fallback}\"; filename*=UTF-8''{quote(filename)}"

File: apps/web/src/shared/excelFilename.ts
/** Tên file Excel.
 *  Lương / OT / KPI / chu kỳ: 08.2026 theo kỳ.
 *  Danh sách NV: 20.08.2026 ngày bấm xuất (giờ VN).
 */

import { todayIsoDateVN } from "./formatDate";

export const COMPANY_EXCEL_SUFFIX = "công ty Dongju Sports VN";

export function excelMonthYearTag(period?: string, ref: Date = new Date()): string {
  const src = (period && period.trim()) || todayIsoDateVN(ref).slice(0, 7);
  const [year, month] = src.split("-");
  if (!year || !month) return src;
  return `${String(Number(month)).padStart(2, "0")}.${year}`;
}

export function excelExportDayTag(ref: Date = new Date()): string {
  const [year, month, day] = todayIsoDateVN(ref).split("-");
  return `${day}.${month}.${year}`;
}

export function companyExcelFilename(
  label: string,
  opts: { period?: string; extra?: string; onExportDay?: boolean; asOf?: Date } = {},
): string {
  const stamp =
    opts.onExportDay || opts.asOf
      ? excelExportDayTag(opts.asOf ?? new Date())
      : excelMonthYearTag(opts.period);
  const parts = [label.trim(), stamp];
  if (opts.extra?.trim()) parts.push(opts.extra.trim());
  parts.push(COMPANY_EXCEL_SUFFIX);
  return `${parts.join(" ")}.xlsx`;
}

export function filenameFromContentDisposition(
  header: string | null | undefined,
  fallback: string,
): string {
  if (!header) return fallback;
  const star = /filename\*=(?:UTF-8''|utf-8'')([^;]+)/i.exec(header);
  if (star) {
    try {
      return decodeURIComponent(star[1].trim().replace(/^"(.*)"$/, "$1"));
    } catch {
      /* dùng fallback / filename= */
    }
  }
  const quoted = /filename="([^"]+)"/.exec(header);
  if (quoted) return quoted[1];
  const plain = /filename=([^;]+)/.exec(header);
  if (plain) return plain[1].trim();
  return fallback;
}
